### PR TITLE
begining of adding file to create new base image for testing script

### PR DESCRIPTION
**Pull request by paetling**
@thieman  WIP do not review yet

### CODE REVIEW

**Comment by paetling** [comment]
@thieman ready for review

**Comment by thieman** [comment]
These functions should all take the client as a parameter, not try to instantiate it themselves. This also makes them way easier to test.

File: dusty/systems/docker/testing_image.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from . import get_docker_client
+
+
+def _ensure_testing_spec_base_image(testing_spec):
+    if 'image' in testing_spec and 'build' in testing_spec:
+        raise RuntimeError('Only 1 of `image` and `build` keys are allowed in testing spec')
```
Comment (thieman) [inline dusty/systems/docker/testing_image.py]: Def need that specs class to consolidate stuff like this, you really don't want validation spread all throughout your app

File: dusty/systems/docker/testing_image.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        docker_client.build(path=testing_spec['build'], tag=image_tag)
+        return image_tag
+    else:
+        raise RuntimeError('One of `image` or `build` is required in testing spec')
+
+def _make_installed_requirements_image(base_image_tag, command, image_name, volumes=[]):
+    docker_client = get_docker_client()
+    container = docker_client.create_container(image=base_image_tag, command=command, volumes=volumes)
+    docker_client.start(container=container['Id'])
+    # new_image = docker_client.commit(container=container['Id'], tag=image_name)
```
Comment (thieman) [inline dusty/systems/docker/testing_image.py]: That's odd. Also your terminology is slipping a bit here. Generally we would use the words `name` and `repository` on an image basically interchangeably, but `tag` always refers to the thing after the colon.

File: dusty/systems/docker/testing_image.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def _ensure_testing_spec_base_image(testing_spec):
+    if 'image' in testing_spec and 'build' in testing_spec:
+        raise RuntimeError('Only 1 of `image` and `build` keys are allowed in testing spec')
+    elif 'image' in testing_spec:
+        return testing_spec['image']
+    elif 'build' in testing_spec:
+        docker_client = get_docker_client()
+        image_tag = 'dusty_testing_base/image'
```
Comment (thieman) [inline dusty/systems/docker/testing_image.py]: Doesn't this need to be unique per app/lib?

File: dusty/systems/docker/testing_image.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    container = docker_client.create_container(image=base_image_tag, command=command, volumes=volumes)
+    docker_client.start(container=container['Id'])
+    # new_image = docker_client.commit(container=container['Id'], tag=image_name)
+    # Above command is not tagging the image, even though it seems like it should be sending
+    # all of the arguments.  Below is a workaround
+    new_image = docker_client.commit(container=container['Id'])
+    docker_client.tag(image=new_image['Id'], repository=image_name, force=True)
+    return image_name
+
+def _make_installed_testing_image(testing_spec, new_image_name, volumes=[]):
```
Comment (thieman) [inline dusty/systems/docker/testing_image.py]: It's probably more clear if we stick to Docker verbs to communicate what we're doing. Probably build then?

File: tests/unit/systems/docker/testing_images_test.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                                                  {'RepoTags': ['dusty/images', 'dusty/image']}]
+        testing_spec = {'once': 'npm install'}
+        new_image_name = 'dusty/image'
+        ensure_image_exists(testing_spec, new_image_name, volumes=['os/path:contianer:path'],force_recreate=True)
+        fake_make_installed_image.assert_has_calls([call(testing_spec, new_image_name, volumes=['os/path:contianer:path'])])
+
+
+
+
+
```
Comment (thieman) [inline tests/unit/systems/docker/testing_images_test.py]: Could you please set up your editor to enforce having exactly one newline at the ends of files? Cuts down on diff noise.

File: tests/unit/systems/docker/testing_images_test.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    @patch('dusty.systems.docker.testing_image._make_installed_testing_image')
+    def test_ensure_image_exists_force(self, fake_make_installed_image, fake_get_docker_client):
+        fake_docker_client = Mock()
+        fake_get_docker_client.return_value = fake_docker_client
+        fake_docker_client.images.return_value = [{'RepoTags': ['dusty', 'dusty/dog']},
+                                                  {'RepoTags': ['dusty/images', 'dusty/image']}]
+        testing_spec = {'once': 'npm install'}
+        new_image_name = 'dusty/image'
+        ensure_image_exists(testing_spec, new_image_name, volumes=['os/path:contianer:path'],force_recreate=True)
+        fake_make_installed_image.assert_has_calls([call(testing_spec, new_image_name, volumes=['os/path:contianer:path'])])
```
Comment (thieman) [inline tests/unit/systems/docker/testing_images_test.py]: M-x replace-string <ret> contianer <ret> container :pig:

**Comment by thieman** [comment]
Seems good in terms of what it's doing and great test coverage. I'm wondering if it makes sense to shore up the specs logic before going any further on writing the tests stuff? What do you think?

File: dusty/systems/docker/testing_image.py (hunk 3 above)
Comment (paetling) [inline dusty/systems/docker/testing_image.py]: this only sticks around for a short period of time, until the new image is build off of this base image.  If this already exists it is overwritten.

File: dusty/systems/docker/testing_image.py (hunk 4 above)
Comment (paetling) [inline dusty/systems/docker/testing_image.py]: i guess in a broad sense this is making an image, so you could call it build.  But there is another function which also makes an image (ensure_testing_spec_base_image)

File: tests/unit/systems/docker/testing_images_test.py (hunk 5 above)
Comment (paetling) [inline tests/unit/systems/docker/testing_images_test.py]: Hmm. I thought I did do that weird.    Will make sure to set it

Comment (paetling) [inline tests/unit/systems/docker/testing_images_test.py]: I set it up to have at least one line, not exactly.  Fixing

File: tests/unit/systems/docker/testing_images_test.py (hunk 6 above)
Comment (paetling) [inline tests/unit/systems/docker/testing_images_test.py]: not sure what this means.

**Comment by paetling** [comment]
I agree it is probably a good idea to clean up the specs repo handling before I do any more test work.

File: tests/unit/systems/docker/testing_images_test.py (hunk 6 above)
Comment (thieman) [inline tests/unit/systems/docker/testing_images_test.py]: You've got the misspelling `contianer` about 8 times

**Pull request merged**